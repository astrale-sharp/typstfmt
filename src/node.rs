use std::fmt::Debug;
use std::rc::Rc;

use typst_syntax::{parse, LinkedNode};

/// This function will give us the root node of our format tree.
///
/// We get rid of unneeded information in LinkedNode and store whatever fundamental information
/// we might need later here.
///
/// The pro of working on a tree is that we can apply transformations to it
///  (see (preserve_pass)[super::preserve_pass]) that will add formatting context.
pub(crate) fn map_tree<'a>(
    node: typst_syntax::LinkedNode<'a>,
    parent: Option<Rc<FmtNode<'a>>>,
) -> FmtNode<'a> {
    use typst_syntax::SyntaxKind::{self, *};

    match node.kind() {
        Raw => FmtKind::Preserve(1).with_text(node, parent),
        Space => FmtKind::Space.with_text(node, parent),
        Linebreak => FmtKind::WithSpacing(Spacing::StrongBrkLine).with_text(node, parent),
        Parbreak => FmtKind::Parbreak.with_text(node, parent),
        Markup => FmtKind::Markup.with_children(node, parent),
        Code => FmtKind::Code.with_children(node, parent),
        CodeBlock => FmtKind::CodeBlock.with_children(node, parent),
        ContentBlock => FmtKind::ContentBlock.with_children(node, parent),
        Math => FmtKind::Math.with_children(node, parent),
        Equation => FmtKind::Equation.with_children(node, parent),
        FuncCall => FmtKind::FuncCall.with_children(node, parent),
        Array | Dict | Args | ListItem | EnumItem | TermItem | Params | Destructuring => {
            FmtKind::ParamsLike.with_children(node, parent)
        }
        Parenthesized => FmtKind::ParamsLikeParenthesized.with_children(node, parent),
        Heading => FmtKind::OneLineMarkup.with_children(node, parent),
        Unary => FmtKind::Unary.with_children(node, parent),
        Binary => FmtKind::Binary.with_children(node, parent),

        LetBinding | ModuleImport | ImportItems | ModuleInclude | SetRule | ShowRule => {
            FmtKind::WithSpacing(Spacing::StrongBrkLine).with_children(node, parent)
        }

        Eof | Spread | Root | Dots | Dot | Star | Underscore | Hashtag | RefMarker | SmartQuote => {
            FmtKind::WithSpacing(Spacing::Destruct).with_text(node, parent)
        }

        LoopBreak
        | LoopContinue
        | FuncReturn
        | Bool
        | Int
        | Float
        | Numeric
        | Str
        | Arrow
        | Not
        | And
        | Or
        | SyntaxKind::None
        | Auto
        | Let
        | Set
        | Show
        | If
        | Else
        | For
        | In
        | While
        | Break
        | Continue
        | Return
        | Import
        | Include
        | As
        | Eq
        | EqEq
        | ExclEq
        | Lt
        | LtEq
        | Gt
        | GtEq
        | PlusEq
        | HyphEq
        | StarEq
        | SlashEq
        | Plus
        | Minus
        | Hat
        | Prime
        | Slash
        | Semicolon
        | Dollar
        | LeftParen
        | RightParen
        | LeftBrace
        | RightBrace
        | LeftBracket
        | RightBracket
        | MathAlignPoint
        | Error
        | Text
        | Shorthand
        | Escape
        | MathIdent => FmtKind::WithSpacing(Spacing::WeakSpace).with_text(node, parent),

        DestructAssignment | WhileLoop | ForLoop | Closure | FieldAccess | Named | Keyed
        | MathFrac | MathRoot | MathAttach | Link | Label | Strong | Emph => {
            FmtKind::WithSpacing(Spacing::WeakSpace).with_children(node, parent)
        }

        MathDelimited | ListMarker | EnumMarker | TermMarker | HeadingMarker | Ref => {
            FmtKind::WithSpacing(Spacing::WeakSpace).with_children(node, parent)
        }

        Ident | Colon | Comma | MathPrimes => {
            FmtKind::WithSpacing(Spacing::StrongSpace).with_text(node, parent)
        }

        Conditional => FmtKind::Conditional.with_children(node, parent),

        BlockComment | LineComment => FmtKind::Comment.with_text(node, parent),
    }
}

#[derive(Clone)]
pub(crate) struct FmtNode<'a> {
    pub(crate) parent: Option<Rc<FmtNode<'a>>>,
    pub(crate) node: typst_syntax::LinkedNode<'a>,
    pub(crate) content: Content<'a>,
    pub(crate) kind: FmtKind,
}

impl<'a> FmtNode<'a> {
    // Todo, we could trust that it's a text and return "" otherwise
    // it should be considered a bug it it doesn't matches Text anyway.
    pub fn text(&self) -> String {
        match &self.content {
            Content::Text(s) => s.to_string(),
            Content::Children(c) => c
                .iter()
                .map(|c| c.text())
                .fold(String::new(), |a, b| format!("{a}{b}")),
        }
    }
}

impl Debug for FmtNode<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}: {:?}\n", self.kind, self.content)
    }
}

#[derive(Clone)]
pub(crate) enum Content<'a> {
    Text(&'a str),
    Children(Vec<FmtNode<'a>>),
}

impl Debug for Content<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Text(arg0) => f.debug_tuple("Text").field(arg0).finish(),
            Self::Children(arg0) => arg0.fmt(f),
        }
    }
}

/// All the kinds that need to be handled differently while formatting

// todo a LetBinding, ForLoop, WhileLoop node forcing a linebreak in code mode or a ; in markup?
#[derive(Debug, Clone, PartialEq)]
pub(crate) enum FmtKind {
    // used only for when in preserve Node, ignored otherwise.
    FuncCall,
    ParamsLike,
    ParamsLikeParenthesized,
    Binary,
    Unary,
    /// The content must be preserved as is.
    ///
    /// This can contain a Raw block or anything targeted in the
    /// (preserve_pass)[super::preserve_pass]
    Preserve(i32),
    Space,
    Parbreak,
    /// Common case that doesn't need precise logic, might be ignored if it is the
    /// child of a custom logic Node such as one with kind : Node.
    WithSpacing(Spacing),

    Markup,
    ContentBlock,
    OneLineMarkup,
    /// if else chain that we must format all at the same time for maximum style.
    Conditional,

    // todo: force line break after loops
    Code,
    CodeBlock,

    Math,
    Equation,

    Comment,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Spacing {
    WeakSpace,
    Destruct,
    StrongSpace,
    StrongBrkLine,
}

impl FmtKind {
    pub(crate) fn with_children<'a>(
        self,
        node: typst_syntax::LinkedNode<'a>,
        parent: Option<Rc<FmtNode<'a>>>,
    ) -> FmtNode<'a> {
        let mut fnode = FmtNode {
            content: Content::Children(vec![]),
            kind: self,
            parent,
            node: node.clone(),
        };

        fnode.content = Content::Children(
            node.children()
                .map(|c| map_tree(c, Some(Rc::new(fnode.clone()))))
                .collect(),
        );

        fnode
    }

    pub(crate) fn with_text<'a>(
        self,
        node: typst_syntax::LinkedNode<'a>,
        parent: Option<Rc<FmtNode<'a>>>,
    ) -> FmtNode<'a> {
        FmtNode {
            content: Content::Text(node.get().text()),
            kind: self,
            parent,
            node,
        }
    }
}

#[test]
fn feature() {
    let s = "#[ text ]";
    let node = parse(s);

    let root = LinkedNode::new(&node);
    let aft = map_tree(root.clone(), None);


    dbg!(&node);
    dbg!(&root);
    dbg!(aft);
}
