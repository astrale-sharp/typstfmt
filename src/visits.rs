// the question is to whom
use crate::{
    node::{Content, FmtKind, FmtNode, Spacing},
    utils::{self, match_first_unwrapped, matches_text},
    writer::Writer,
};

use typst_syntax::ast::*;

/// respects everything but spaces
fn visit(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    match &node.kind {
        FmtKind::Space => (),
        FmtKind::FuncCall => visit_func_call(node, w),
        FmtKind::ParamsLike => visit_params_like(node, w, false),
        FmtKind::ParamsLikeParenthesized => visit_params_like(node, w, true),
        FmtKind::Binary => visit_binary(node, w),
        FmtKind::Unary => visit_unary(node, w),
        FmtKind::Preserve(_) => visit_preserve(node, w),
        FmtKind::Parbreak => visit_parbreak(node, w),
        FmtKind::WithSpacing(s) => visit_spacing(s, node, w),
        FmtKind::Markup => visit_markup(node, w, true),
        FmtKind::ContentBlock => visit_content_block(node, w),
        FmtKind::OneLineMarkup => visit_one_line_markup(node, w),
        FmtKind::Conditional => visit_conditional(node, w),
        FmtKind::Code => visit_code(node, w),
        FmtKind::CodeBlock => visit_code_block(node, w),
        FmtKind::Math => visit_math(node, w),
        FmtKind::Equation => visit_equation(node, w),
        FmtKind::Comment => visit_comment(node, w),
    }
}

fn visit_comment(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    visit_preserve(node, w);
    w.push_str("\n")
}

/// Todo, deals with most common cases that don't need special logic.
///
/// Might not always be called on (Node)[FmtNode] of kind [FmtKind::WithSpacing] if their in
/// another visit function that has it's own logic.
pub(crate) fn visit_spacing(s: &Spacing, node: &FmtNode<'_>, w: &mut Writer<'_>) {
    w.push_str(&node.text().trim());
    match s {
        Spacing::WeakSpace => w.push_str(" "),
        Spacing::Destruct => (),
        Spacing::StrongSpace => w.push_str(" "),
        Spacing::StrongBrkLine => w.push_str("\n"),
    }
}

pub(crate) fn visit_spacing_line_wrapped(s: &Spacing, node: &FmtNode<'_>, w: &mut Writer<'_>) {
    w.push_str_with_limit(&node.text().trim());
    match s {
        Spacing::WeakSpace => w.push_str(" "),
        Spacing::Destruct => (),
        Spacing::StrongSpace => w.push_str(" "),
        Spacing::StrongBrkLine => w.push_str("\n"),
    }
}

/// simply push respecting everything and tag it as preserve.
/// todo: can we rely on node.text? we should have a recursive use of visit preserve instead.
pub(crate) fn visit_preserve(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    w.mark_preserve();
    w.push_str(&node.text());
    w.mark_stop_preserve();
}

pub(crate) fn visit_content_block(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    let mark = w.get_mark();
    visit_content_block_with(node, w, true);
    // todo or go above max length
    if w.buffer[mark..].contains("\n") {
        w.rewind(mark);
        visit_content_block_with(node, w, false);
    }
}

/// todo: respect absence of space just after [ and just before ]
/// - try a tight and expand mode.
pub(crate) fn visit_content_block_with(node: &FmtNode<'_>, w: &mut Writer<'_>, tight: bool) {
    let Content::Children(c) = &node.content else {
        unreachable!()
    };
    // todo rm this clone
    assert!(matches_text(c.last(), "["));
    assert!(matches_text(c.first(), "]"));

    w.push_str("[");
    w.mark_indent();
    let mut c = c.iter();
    c.next();
    if let Some(c) = c.next() {
        debug_assert_eq!(c.kind, FmtKind::Markup);
        visit_markup(node, w, tight)
    } else {
        debug_assert!(false);
    }
    w.mark_dedent();
    w.push_str("]");
}

/// if config.wrap_text Wrap text similar to visit_basic but checking max line length
/// else respect space but replace " "+ -> " " and <whitespace> "\n"+ <whitespace> -> "\n"
// optimize, return a bool to fail tight in advance
pub(crate) fn visit_markup(node: &FmtNode<'_>, w: &mut Writer<'_>, tight: bool) {
    let Content::Children(c) = &node.content else {
        unreachable!()
    };
    let start_space = utils::next_is_space(c.iter());
    let end_space = utils::next_is_space(c.iter().rev());

    let c = &mut c.iter().filter(|c| c.kind != FmtKind::Space).peekable();
    let space_kind = if tight { " " } else { "\n" };
    if start_space {
        w.push_str(space_kind)
    }
    for node in c {
        match &node.kind {
            FmtKind::Space if w.config.line_wrap => (),
            FmtKind::WithSpacing(s) if w.config.line_wrap => visit_spacing_line_wrapped(s, node, w),
            FmtKind::Space => {
                if node.text().contains("\n") {
                    w.push_str("\n")
                }
            }
            FmtKind::ContentBlock if tight => visit_content_block_with(node, w, tight),
            _ => visit(node, w),
        }
    }
    if end_space {
        w.push_str(space_kind)
    }
}

pub(crate) fn visit_code(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    todo!()
}

pub(crate) fn visit_code_block(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    let Content::Children(c) = &node.content else {
        unreachable!()
    };
    let c = &mut c.iter().filter(|c| c.kind != FmtKind::Space);
    assert!(matches_text(c.next(), "{"));
    assert!(matches_text(c.next_back(), "}"));
    // todo put a mark try one line code else rewind

    w.push_str("{\n");
    w.mark_indent();
    for node in c {
        visit(node, w)
    }
    w.mark_dedent();
    w.push_str("\n}");
}

pub(crate) fn visit_math(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    todo!()
}

pub(crate) fn visit_equation(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    let Content::Children(c) = &node.content else {
        unreachable!()
    };
    let c = &mut c.iter().filter(|c| c.kind != FmtKind::Space);
    assert!(matches_text(c.next(), "$\n"));
    assert!(matches_text(c.next_back(), "$\n"));

    w.push_str("$\n");
    w.mark_indent();
    visit_math(&match_first_unwrapped::<Math>(c), w);
    w.mark_dedent();
    w.push_str("\n$");
}

pub(crate) fn visit_parbreak(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    w.push_str("\n\n")
}

pub(crate) fn visit_conditional(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    todo!()
}

pub(crate) fn visit_one_line_markup(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    todo!()
}

pub(crate) fn visit_unary(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    todo!()
}

pub(crate) fn visit_binary(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    todo!()
}

pub(crate) fn visit_params_like(node: &FmtNode<'_>, w: &mut Writer<'_>, parenthesized: bool) {
    todo!()
}

pub(crate) fn visit_func_call(node: &FmtNode<'_>, w: &mut Writer<'_>) {
    todo!()
}
