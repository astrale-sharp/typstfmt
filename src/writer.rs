use crate::Config;
use itertools::Itertools;
use std::{
    collections::{HashMap, HashSet},
    mem,
};

// TODO, not a priority 
// remove post process indents, instead we want 2 functions on the writer,
// one that respects indentation and one that doesn't for Preserve NodeFmts.
// This should come with indent and dedent increasing an index in ctx, used to give indentation

/// Writer is used to write your formatted output.
///
/// It comes with the following features :
/// - Markers: you place a mark by calling [Writer::mark], you can use this mark
/// to jump back and redo your formatting should you see it didn't respect some rules.
/// - Todo: rewinding, go back to a position, removing all markers that were introduced after
/// as well as resetting the result to it's past state.
/// Example :
/// ```ignore
/// fn visit_params(/* */) {
///     let mark = self.mark();
///     visit_params_tight(self);
///     if !line_is_too_long(ctx.string_after_mark(mark)) {
///         // we're done yay
///         return
///     } else {
///         ctx.rewind(mark);
///         visit_params_long();
///     }
/// }
/// ```
/// - Indent, Dedent, Preserve: You must specify where you want indent to start and end.
/// It will be applied as a later step respecting the Preserve markers where indentation is
/// not applied (Raw, fmt::off, etc.)
pub(crate) struct Writer<'a> {
    pub(crate) config: Config,
    pub(crate) buffer: &'a mut String,
    pub(crate) marks: Vec<Mark>,
}

#[derive(Debug, PartialEq, Clone)]
pub(crate) struct Mark {
    kind: MarkKind,
    pos: usize,
}
#[derive(Debug, PartialEq, Clone)]
pub(crate) enum MarkKind {
    Indent,
    Dedent,
    Preserve,
    StopPreserve,
}

impl MarkKind {
    pub(crate) fn to_mark(self, pos: usize) -> Mark {
        Mark { kind: self, pos }
    }
}

impl<'a> Writer<'a> {
    pub(crate) fn new(config: Config, buffer: &'a mut String) -> Self {
        Self {
            config,
            buffer,
            marks: vec![],
        }
    }

    pub fn get_mark(&self) -> usize {
        self.buffer.len()
    }

    pub fn rewind(&mut self, mark: usize) {
        *self.buffer = self.buffer[..mark].to_string();
        while let Some(x) = self.marks.last() {
            if x.pos < mark {
                self.marks.pop();
            }
        }
    }

    // needed when post process adds indentation level.
    pub fn mark_indent(&mut self) {
        self.marks.push(MarkKind::Indent.to_mark(self.buffer.len()))
    }

    pub fn mark_dedent(&mut self) {
        self.marks.push(MarkKind::Dedent.to_mark(self.buffer.len()))
    }

    pub fn mark_preserve(&mut self) {
        self.marks
            .push(MarkKind::Preserve.to_mark(self.buffer.len()))
    }

    pub fn mark_stop_preserve(&mut self) {
        self.marks
            .push(MarkKind::StopPreserve.to_mark(self.buffer.len()))
    }

    /// TODO, check the indexes with BASIC tests
    pub fn post_process_indents(&mut self) {
        let lines = self.buffer.split_inclusive('\n');
        let lines_len = lines.clone().count();
        let sizes = lines.clone().map(|s| 0..str::len(s)).collect_vec();
        // (line number -> bytes idx)
        let sizes = (1..=sizes.len())
            .map(|x| {
                sizes.iter().take(x).fold(0..0, |init, next| {
                    init.end..(init.end + next.end - next.start)
                })
            })
            .collect_vec();
        let line_at_pos = |i: &Mark| (&sizes).iter().position(|range| range.contains(&i.pos));

        // We'll match closed (except typstfmt::off) pairs of marks
        let mut marks = mem::take(&mut self.marks);
        marks.sort_by(|x, y| x.pos.cmp(&y.pos));
        debug_assert!(
            marks
                .iter()
                .map(|x| x.pos)
                .position_min()
                .unwrap_or_default()
                == 0
        );

        let ident_dedent_pairs = get_matching_pairs(&marks, MarkKind::Indent, MarkKind::Dedent);
        let preserve_pairs = get_matching_pairs(&marks, MarkKind::Preserve, MarkKind::StopPreserve);
        dbg!(&preserve_pairs);
        // - A (line number -> is_preserved) register.
        let mut preserved_lines = HashSet::new();
        for pair in preserve_pairs {
            let start = line_at_pos(pair.0).unwrap();
            // it's okay if `//typstfmt::off`` is not closed.
            let end = line_at_pos(pair.1).unwrap(); //_or(lines_len);
            for k in start..end {
                dbg!(k);
                preserved_lines.insert(k);
            }
        }

        let ident_line_range = ident_dedent_pairs
            .into_iter()
            .map(|(i, d)| line_at_pos(i).unwrap()..line_at_pos(d).unwrap())
            .collect_vec();
        // - A (line number -> indent level) HashMap.
        let mut line_to_ident = HashMap::new();
        for r in ident_line_range {
            // we skip the line where the mark appeared cause we only add indentation once we've
            // added a line break.
            for k in r.skip(1) {
                line_to_ident.entry(k).and_modify(|x| *x += 1).or_insert(1);
            }
        }

        let mut res = String::new();
        for (idx, line) in lines.enumerate() {
            if let Some(i) = line_to_ident.get(&idx) {
                if !preserved_lines.contains(&idx) {
                    // todo, config for tabs
                    res.push_str(&self.config.indent.get(*i))
                }
            }
            res.push_str(line);
        }
        *self.buffer = res;
    }

    pub(crate) fn push_str(&mut self, s: &str) {
        // TODO if we remove post process indent
        // check if we added a line feed and add indentation accordingly
        self.buffer.push_str(s)
    }

    /// pushes the thing checking for overflowing the line first and breaking if needed
    ///
    /// returns true if it broke the line
    pub(crate) fn push_str_with_limit(&mut self, s: &str) {
        if s == "" {
            return;
        }

        if self.last_line_length() + s.len() >= self.config.max_line_length {
            self.new_line()
        }
        self.buffer.push_str(s);
    }

    pub(crate) fn new_line(&mut self) {
        self.buffer.push('\n')
    }

    pub(crate) fn space(&mut self) {
        self.buffer.push(' ')
    }

    pub(crate) fn last_line_length(&self) -> usize {
        use unicode_segmentation::UnicodeSegmentation;
        self.buffer
            .split('\n')
            .last()
            .unwrap_or("")
            .trim()
            .graphemes(true)
            .count()
    }
}

fn get_matching_pairs(
    marks: &Vec<Mark>,
    opening_kind: MarkKind,
    closing_kind: MarkKind,
) -> Vec<(&Mark, &Mark)> {
    let mut res = vec![];
    for (i, m) in marks.iter().enumerate() {
        if opening_kind == m.kind {
            let mut c = 0;
            for k in (i + 1)..marks.len() {
                let m2 = &marks[k];
                if m2.kind == opening_kind {
                    c += 1;
                } else if m2.kind == closing_kind {
                    if c == 0 {
                        res.push((m, m2));
                        break;
                    } else {
                        c -= 1;
                    }
                }
            }
        }
    }
    res
}

#[test]
fn test_get_pairs() {
    let mi1 = "\n#[".len();
    let mi2 = "\n#[\ntext #[".len();
    let md2 = "\n#[\ntext #[\ntext\n".len();
    let md1 = "\n#[\ntext #[\ntext\n]\n".len();
    let preserve = "\n#[\ntext #[\n".len();
    let stop_preserve = "\n#[\ntext #[\ntext".len();

    let mut binding = String::new();
    let mut w = Writer::new(Config::default(), &mut binding);

    w.marks = vec![
        MarkKind::Indent.to_mark(mi1),
        MarkKind::Indent.to_mark(mi2),
        MarkKind::Dedent.to_mark(md2),
        MarkKind::Dedent.to_mark(md1),
        MarkKind::Preserve.to_mark(preserve),
        MarkKind::StopPreserve.to_mark(stop_preserve),
    ];

    let res = get_matching_pairs(&w.marks, MarkKind::Indent, MarkKind::Dedent);
    assert!(
        res == vec![
            (
                &MarkKind::Indent.to_mark(mi1),
                &MarkKind::Dedent.to_mark(md1)
            ),
            (
                &MarkKind::Indent.to_mark(mi2),
                &MarkKind::Dedent.to_mark(md2)
            )
        ]
    );
    let res = get_matching_pairs(&w.marks, MarkKind::Preserve, MarkKind::StopPreserve);
    dbg!(&res);
    assert!(
        res == vec![(
            &MarkKind::Preserve.to_mark(preserve),
            &MarkKind::StopPreserve.to_mark(stop_preserve)
        )]
    );
}
