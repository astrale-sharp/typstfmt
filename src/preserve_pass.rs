use super::node::{Content, FmtKind, FmtNode};

/// Matches `// typstfmt::off` and `// typstfmt::on` Typst's comments
/// and Mark as Preserved all the nodes that are englobed. They will be preserved from
/// modifications during formatting.
///
/// Note that if one of your children has an off command, it doesn't ensure that the on
/// command is in one of your other children or exists at all.
///
/// Example:
/// In the next snippet, everything between the two commands is preserved.
/// ```ignore
/// f(
///     // t::off
///     g([text  text  text]),2,
/// )
/// lorem ipsum
/// // t::on
/// ```
///
/// Note however that starting a node preserved will not give you the option to change your mind halfway, as
/// all the logic of handling a node happens in the beginning of it's formatting process.
///
/// Hence in the next snippet, the whole f function call will be preserved (but not the lorem ispum)
/// ```ignore
/// // t::off
/// f(
///     // t::on
///     g([text  text  text]),2,
/// )
/// lorem ipsum
/// ```
///
/// TODO: TEST lorem ispum is not preserved.
pub(crate) fn preserve_pass(node: &mut FmtNode) -> PreserveData {
    match node.kind {
        FmtKind::Comment => {
            let text = node.text();
            if text.starts_with("// typstfmt::off") {
                return PreserveData::new(1, false);
            } else if text.starts_with("// typstfmt::on") {
                return PreserveData::new(-1, false);
            } else if text.contains("typstfmt::") {
                // TODO WARN HERE
            }
        }
        _ => ()
    }

    match &mut node.content {
        Content::Text(_) => PreserveData::default(),
        Content::Children(c) => {
            let mut node_data = PreserveData::new(0, true);
            c.iter_mut().for_each(|c| {
                let child_data = preserve_pass(c);
                // we ignore typstfmt::on if we weren't preserving already
                node_data.preserve_idx = (node_data.preserve_idx + child_data.preserve_idx).max(0);

                // if one of the children started the preservation, it handled itself already.
                if node_data.preserve_idx > 0 {
                    // We had things to do so we might have already handles things correctly
                    node_data.parent_must_handle = false;
                    match &mut c.content {
                        Content::Text(_) => c.kind = FmtKind::Preserve(node_data.preserve_idx),
                        Content::Children(_) => {
                            if child_data.parent_must_handle || node_data.preserve_idx > 1 {
                                c.kind = FmtKind::Preserve(node_data.preserve_idx)
                            }
                            // else the node handled itself already
                        }
                    }
                }
            });
            node_data
        }
    }
}

/// preserve index and parent must handle me flags
#[derive(Default)]
pub(crate) struct PreserveData {
    preserve_idx: i32,
    parent_must_handle: bool,
}

impl PreserveData {
    fn new(preserve_idx: i32, parent_must_handle: bool) -> Self {
        Self {
            preserve_idx,
            parent_must_handle,
        }
    }
}



macro_rules! tree_has_node_matching {
    (
        $test_name:ident,
        $snippet:expr,
        $kind_fn:expr,
        $parent_fn:expr,
        $($matching:tt)*
    ) => {
        #[test]
        fn $test_name() {
            let snippet = typst_syntax::parse($snippet);
            let root = typst_syntax::LinkedNode::new(&snippet);
            let tree = crate::map_tree(root, None);
            dbg!(&tree);
            let parent : &dyn Fn(Option<&FmtNode>) -> bool = $parent_fn;
            let kind : &dyn Fn(&FmtKind) -> bool = &$kind_fn;
            assert!(
                contains_matching(&tree, &|node| {
                    if !(parent(node.parent.as_deref()) && kind(&node.kind))  {
                        return false
                    }

                    match &node.content {
                        Content::Children(c) => match c.as_slice() {
                            $($matching)* => true,
                            _ => false,
                        },
                        _ => false,
                    }})
                );
            }
        };
        (
            $test_name:ident,
            $snippet:expr,
            $kind_fn:expr,
            $($matching:tt)*
        ) => {
            tree_has_node_matching!(
                $test_name,
                $snippet,
                $kind_fn,
                &|_|true,
                $($matching)*
            );
        };
}

pub fn contains_matching(r: &FmtNode, condition: &impl Fn(&FmtNode) -> bool) -> bool {
    dbg!(r);
    condition(r)
        || if let Content::Children(c) = &r.content {
            !c.is_empty() && c.iter().any(|x| contains_matching(x, condition))
        } else {
            false
        }
}

tree_has_node_matching!(
    basic_with_parent,
    "#[ // typstfmt::off\ntext\n// typstfmt::on\n]text",
    |kind| matches!(kind, FmtKind::Markup),
    x if x.iter().find(|c|c.kind == FmtKind::Comment).is_some()
);

tree_has_node_matching!(
    basic,
    "#[ // typstfmt::off\ntext\n// typstfmt::on\n]text",
    |_| true,
    x if x.iter().find(|c|&c.kind == &FmtKind::FuncCall).is_some()
);
