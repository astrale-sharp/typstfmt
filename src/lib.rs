#![doc = include_str!("../README.md")]
// #![allow(unused)]
// #![allow(warnings)]
#![warn(
    clippy::all,
    clippy::pedantic,
    clippy::print_stdout,
    clippy::print_stderr,
    clippy::dbg_macro
)]

mod config;
/// Our format tree, that we will visit in order to format the code.
mod node;
/// Special operation applied to our tree to take care of giving commands via
/// Typst's comments.
mod preserve_pass;
mod utils;
/// Here lies the formatting logic
mod visits;
/// Handles writing to an output, indentation as a post_process,
/// rewinding if things some condition was not respected.
mod writer;

pub use config::Config;
use node::map_tree;
use preserve_pass::preserve_pass;
use visits::*;
use writer::Writer;
// mod tests;

use typst_syntax::{parse, LinkedNode};

/// we visit our [FmtNode] tree, using the [Writer] to specify how we want
/// our formatting to be done.
///
/// Then we apply indentation as post processing.
#[must_use]
pub fn format(s: &str, config: Config) -> String {
    let init = parse(s);
    let mut s = String::new();
    let mut writer = Writer::new(config, &mut s);
    let root = LinkedNode::new(&init);
    let mut root = map_tree(root, None);
    let _ = preserve_pass(&mut root);
    visit_markup(&root, &mut writer, false);
    writer.post_process_indents();
    regex::Regex::new("( )+\n")
        .unwrap()
        .replace_all(&s, "\n")
        .to_string()
}

/* very basic tests */

#[test]
fn test_indent() {
    let mut snippet = r#"
#[
text #[
text
]
]"#
    .to_string();

    let expected = r#"
#[
  text #[
    text
  ]
]"#;

    let mi1 = "\n#[".len();
    let mi2 = "\n#[\ntext #[".len();
    let md2 = "\n#[\ntext #[\ntext\n".len();
    let md1 = "\n#[\ntext #[\ntext\n]\n".len();

    let mut w = Writer::new(Config::default(), &mut snippet);
    w.marks = vec![
        MarkKind::Indent.to_mark(mi1),
        MarkKind::Indent.to_mark(mi2),
        MarkKind::Dedent.to_mark(md2),
        MarkKind::Dedent.to_mark(md1),
    ];
    w.post_process_indents();
    println!("snippet:");
    println!("{snippet}");
    println!("snippet:?");
    println!("{snippet:?}");

    assert!(snippet == expected);
}

#[cfg(test)]
use crate::writer::MarkKind;

#[test]
fn test_preserve() {
    let mut snippet = r#"
#[
text #[
text
]
]"#
    .to_string();
    println!("init: {snippet}");

    let expected = r#"
#[
  text #[
text
  ]
]"#;

    let mi1 = "\n#[".len();
    let mi2 = "\n#[\ntext #[".len();
    let md2 = "\n#[\ntext #[\ntext\n".len();
    let md1 = "\n#[\ntext #[\ntext\n]\n".len();
    let preserve = "\n#[\ntext #[\n".len();
    let stop_preserve = "\n#[\ntext #[\ntext\n".len();

    let mut w = Writer::new(Config::default(), &mut snippet);
    w.marks = vec![
        MarkKind::Indent.to_mark(mi1),
        MarkKind::Indent.to_mark(mi2),
        MarkKind::Dedent.to_mark(md2),
        MarkKind::Dedent.to_mark(md1),
        MarkKind::Preserve.to_mark(preserve),
        MarkKind::StopPreserve.to_mark(stop_preserve),
    ];
    w.post_process_indents();
    println!("fmt : {snippet}");
    println!("expe: {expected}");

    assert!(snippet == expected);
}

#[test]
fn test_preserve_pass() {
    let snippets = [
        r#"notp // typstfmt::off
#f(x : [p]) 
// typstfmt::on
notp "#,
        r#"#f([notp], // typstfmt::off
)
// typstfmt::off
// typstfmt::on
p
// typstfmt::on
notp"#,
        r#"
// typstfmt::off
#[
// typstfmt::off
#[
// typstfmt::off
#[
// typstfmt::off
text
]
]
]
// typstfmt::on
// typstfmt::on
// typstfmt::on
text
// typstfmt::on
text
"#,
    ];

    for snippet in snippets {
        let parse = parse(snippet);
        let root = LinkedNode::new(&parse);
        println!("===");
        println!("{snippet}");
        println!("===");
        println!("{:?}", root);
        let mut root = map_tree(root, None);
        println!("===");
        println!("{:?}", root);
        println!("===");
        preserve_pass(&mut root);
        println!("{:?}", root);
        println!("===");
        insta::assert_debug_snapshot!(root)
    }
}
