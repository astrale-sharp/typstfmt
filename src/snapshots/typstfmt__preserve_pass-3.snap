---
source: src/lib.rs
expression: root
---
Markup: [Space: Text("\n"), Preserve(1): Text("// typstfmt::off"), Preserve(1): Text("\n"), Preserve(1): Text("#"), Preserve(4): [WithSpacing(WeakSpace): Text("["), Preserve(3): [Space: Text("\n"), Preserve(1): Text("// typstfmt::off"), Preserve(1): Text("\n"), Preserve(1): Text("#"), Preserve(3): [WithSpacing(WeakSpace): Text("["), Preserve(2): [Space: Text("\n"), Preserve(1): Text("// typstfmt::off"), Preserve(1): Text("\n"), Preserve(1): Text("#"), Preserve(2): [WithSpacing(WeakSpace): Text("["), Markup: [Space: Text("\n"), Preserve(1): Text("// typstfmt::off"), Preserve(1): Text("\n"), Preserve(1): Text("text"), Preserve(1): Text("\n")], Preserve(1): Text("]")], Preserve(2): Text("\n")], Preserve(2): Text("]")], Preserve(3): Text("\n")], Preserve(3): Text("]")], Preserve(4): Text("\n"), Preserve(3): Text("// typstfmt::on"), Preserve(3): Text("\n"), Preserve(2): Text("// typstfmt::on"), Preserve(2): Text("\n"), Preserve(1): Text("// typstfmt::on"), Preserve(1): Text("\n"), Preserve(1): Text("text"), Preserve(1): Text("\n"), Comment: Text("// typstfmt::on"), Space: Text("\n"), WithSpacing(WeakSpace): Text("text"), Space: Text("\n")]
